x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/IndicesAccessControl.java

```diff
@@ -82,20 +84,56 @@ public FieldPermissions getFieldPermissions() {
          *         then this means that there are no document level restrictions
          */
         @Nullable
-        public Set<BytesReference> getQueries() {
-            return queries;
+        public DocumentPermissions getDocumentPermissions() {
+            return documentPermissions;
+        }
+
+        public static IndexAccessControl scopedIndexAccessControl(IndexAccessControl indexAccessControl,
+                IndexAccessControl scopedByIndexAccessControl) {
+            final boolean granted;
+            if (indexAccessControl.granted == indexAccessControl.granted) {
+                granted = indexAccessControl.granted;
+            } else {
+                granted = false;
+            }
+            FieldPermissions fieldPermissions = FieldPermissions.scopedFieldPermissions(indexAccessControl.fieldPermissions,
+                    scopedByIndexAccessControl.fieldPermissions);
+            DocumentPermissions documentPermissions = DocumentPermissions.scopedDocumentPermissions(
+                    indexAccessControl.getDocumentPermissions(), scopedByIndexAccessControl.getDocumentPermissions());
+            return new IndexAccessControl(granted, fieldPermissions, documentPermissions);
         }
 
         @Override
         public String toString() {
             return "IndexAccessControl{" +
                     "granted=" + granted +
                     ", fieldPermissions=" + fieldPermissions +
-                    ", queries=" + queries +
+                    ", documentPermissions=" + documentPermissions +
                     '}';
         }
     }
 
+    public static IndicesAccessControl scopedIndicesAccessControl(IndicesAccessControl indicesAccessControl,
+            IndicesAccessControl scopedByIndicesAccessControl) {
+        final boolean granted;
+        if (indicesAccessControl.granted == scopedByIndicesAccessControl.granted) {
+            granted = indicesAccessControl.granted;
+        } else {
+            granted = false;
+        }
+        Set<String> indexes = indicesAccessControl.indexPermissions.keySet();
+        Set<String> otherIndexes = scopedByIndicesAccessControl.indexPermissions.keySet();
+        Set<String> commonIndexes = Sets.intersection(indexes, otherIndexes);
```

Comment:
This doesn't seem to take into account wildcards, which seems like a big limitation...

I have access to `*` indices, can't I create an API-key that is limited to`logstash-*`? Or perhaps I'm missing something (I'm still trying to get my head around the big picture).

Reply:
Ah, yes.
I can't help but feel that we have too many classes that deal with rights-to-do-things-to-one-or-more-indices. I have no solution, but it's gotten to the point where they won't all stay in my brain at the same time. 🤕